### PR TITLE
CONSOLE-3792: Add extension point to enable customized create project modal for dynamic plugins

### PR DESCRIPTION
**Pull request by sg00dwin**
…

### CODE REVIEW

**Comment by TheRealJon** [comment]
I just realized that we probably need to change the existing `createNamespaceModal,` `createProjectModal,` and `createNamespaceOrProjectModal` functions to use the `useModal` hook from the dynamic plugin SDK. The `createModalLauncher` function, part of our internal app, passes a different set of props than the launcher returned by `useModal`. We will want the new extension point to use props from the stable SDK API, not from our internal app.

**Comment by TheRealJon** [comment]
> I just realized that we probably need to change the existing `createNamespaceModal,` `createProjectModal,` and `createNamespaceOrProjectModal` functions to use the `useModal` hook from the dynamic plugin SDK. The `createModalLauncher` function, part of our internal app, passes a different set of props than the launcher returned by `useModal`. We will want the new extension point to use props from the stable SDK API, not from our internal app.

This also means that these functions will need to be migrated to hooks

File: frontend/packages/console-shared/src/components/modals/CreateProjectModal.tsx [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+import { FLAGS } from '@console/shared';
+import { ModalComponent } from 'packages/console-dynamic-plugin-sdk/src/app/modal-support/ModalProvider';
+
+const DefaultCreateProjectModal: ModalComponent<CreateProjectModalProps> = ({
+  closeModal,
+  onSubmit,
+}) => {
+  const navigate = useNavigate();
+  const dispatch = useDispatch();
+  const [inProgress, setInProgress] = React.useState(false);
```
Comment (jhadvig) [inline frontend/packages/console-shared/src/components/modals/CreateProjectModal.tsx:35]: Lets type these as well

**Review by jhadvig — COMMENTED** [review]
Thanks for this change @sg00dwin 👍 
Adding couple of smaller comments.
Also it looks you need to regenerate the i18n.

File: dynamic-demo-plugin/package.json [hunk 2]
```diff
@@ -71,7 +71,8 @@
       "projectOverviewInventory": "./components/ProjectOverview/Inventory",
       "projectOverview": "./utils/project-overview",
       "modalPage": "./components/Modals/ModalPage",
-      "sampleErrorBoundaryPage": "./components/ErrorBoundary/SampleErrorBoundaryPage"
+      "sampleErrorBoundaryPage": "./components/ErrorBoundary/SampleErrorBoundaryPage",
+      "CreateProjectModal": "./components/Modals/CreateProjectModal"
```
[the remark this answers is not in the document]
Comment (sg00dwin) [reply dynamic-demo-plugin/package.json]: updated

**Comment by sg00dwin** [comment]
QE Approver
/assign @yapei

Docs Approver:
/assign @opayne1

PX Approver:
/assign @RickJWagner

File: frontend/packages/console-shared/src/components/modals/CreateProjectModal.tsx (hunk 1 above)
Comment (TheRealJon) [reply frontend/packages/console-shared/src/components/modals/CreateProjectModal.tsx:35]: Type is inferred from the initial value for primitive values.

**Review by TheRealJon — APPROVED** [review]
/lgtm

**Comment by TheRealJon** [comment]
/lgtm cancel

Since I helped implement this, I'll let someone else tag it.

File: frontend/public/components/factory/modal.tsx [hunk 3]
```diff
@@ -12,6 +12,7 @@ import store from '../../redux';
 import { ButtonBar } from '../utils/button-bar';
 import { history } from '../utils/router';
 
+/** @deprecated Use dynamic plugin sdk 'useModal' hook instead */
```
Comment (rhamilto) [inline frontend/public/components/factory/modal.tsx:15]: 👍

**Review by rhamilto — APPROVED** [review]
/lgtm

Nice work!

**Comment by yapei** [comment]
tested following scenarios:
1. verify `Create Project` modal is using customized modal on several pages(project lists page; getting started page; project dropdown)
2. check `Create Namespace` modal, there is some problem here, user is unable to open the `Default network policy` dropdown
Screenshot 2024-06-12 at 2 20 54 PM

**Comment by yapei** [comment]
also one more question: customized namespace creation modal is not supported/implemented in this PR, right?

**Comment by sg00dwin** [comment]
> also one more question: customized namespace creation modal is not supported/implemented in this PR, right? Do we plan to support it in future?

That is not implemented with this PR and there is no plans to support it.

**Comment by yapei** [comment]
`Create Namespace` modal still have some issues
- user need click twice to successfully switch to `Deny all inbound traffic` 
- the dropdown item changed back to `No restrictions` after user clicks on `Create` button

https://github.com/openshift/console/assets/12692381/d6cd7f7b-96f2-4b34-84d3-e87993d9ef7f

**Comment by sg00dwin** [comment]
tc-approved

**Comment by sg00dwin** [comment]
Hi @yapei I've pushed updates for those issues.

**Comment by yapei** [comment]
1. `Create Namespace` modal showing `Default network policy` with text `allow` 
Screenshot 2024-06-24 at 11 28 40 AM
2. When creating a Namespace with `Deny all inbound traffic` , no NetworkPolicy is created while default-deny networkpolicy should be created

**Comment by sg00dwin** [comment]
I have pushed further updates for those issues. Please review. 
namespace-select

**Comment by yapei** [comment]
@sg00dwin thanks for all the updates, now everything works fine
/label qe-approved

**Review by jhadvig — APPROVED** [review]
Really nice job @sg00dwin 👍 
/lgtm

**Pull request merged**